python/tvm/contrib/target/onnx.py

```diff
@@ -0,0 +1,752 @@
+
+    def _add_params(self, node_entry, idx):
+        """Add param value to initializer and name to inputs"""
+        param_name = node_entry['name']
+        assert param_name in self._params, "The parameter {0} is not present" \
+                                           "in params dict provided.".format(param_name)
+        value = self._params[param_name]
+        numpy_array = value.asnumpy()
+        tensor = numpy_helper.from_array(numpy_array, param_name)
+        self._mc.add_initializers([tensor])
+        dtype = onnx.mapping.NP_TYPE_TO_TENSOR_TYPE[numpy_array.dtype]
+        input = onnx.helper.make_tensor_value_info(param_name,
+                                                   dtype,
+                                                   shape=numpy_array.shape)
+        self._mc.add_inputs([input])
+
+    def _add_constant_input(self, node_entry, idx):
+        """Create named input for constant and add it to container inputs.
+        If input is a parameter then add to param
+        """
+        node = node_entry['node']
+        if not node_entry['name']:
+            node_entry['name'] = self._tuple_to_name([idx, 0, 0])
+        param_name = node_entry['name']
+        self._params[param_name] = node.data
+        self._add_params(node_entry, idx)
+
+    def _add_input(self, node_entry, idx):
+        """Add input node to container inputs. If input is a parameter then add to param"""
+        if node_entry['name'] in self._params:
+            self._add_params(node_entry, idx)
+        else:
+            type = node_entry['types'][0]
+            dtype = onnx.mapping.NP_TYPE_TO_TENSOR_TYPE[numpy.dtype(type.dtype)]
+            input = onnx.helper.make_tensor_value_info(node_entry['name'],
+                                                       dtype,
+                                                       shape=type.concrete_shape)
+            self._mc.add_inputs([input])
+
+    def _add_output(self, node_entry, idx):
+        """Add output node to container outputs."""
+
+        type = node_entry['types'][0]
+        dtype = onnx.mapping.NP_TYPE_TO_TENSOR_TYPE[numpy.dtype(type.dtype)]
+        output = onnx.helper.make_tensor_value_info(self._tuple_to_name([idx, 0, 0]),
+                                                    dtype,
+                                                    shape=type.concrete_shape)
+        self._mc.add_outputs([output])
+
+
+def to_onnx(relay_ir, params, name, opset_version=11, path=None):
+    """Convert a Relay Function Module into an equivalent ONNX and serialize it to the path
+
+    Parameters
+    ----------
+    relay_ir : tvm.ir.IRModule or tvm.relay.Function
+        The relay module object
+
+    params : dict
+        dict of the parameter names and NDarray values
+
+    path : str
+        The path where ONNX model will be saved
+
+    Returns
+    -------
+    inferred_model : tvm.relay.Module
+        The relay module
```

Comment:
I think the return value is not relay.Module but onnx.ModelProto.

Reply:
Done